datafusion/physical-expr/src/array_expressions.rs

```diff
@@ -2150,6 +2150,61 @@ pub fn array_has_all(args: &[ArrayRef]) -> Result<ArrayRef> {
     Ok(Arc::new(boolean_builder.finish()))
 }
 
+macro_rules! array_sum_internal(
+    ($ARRAY:expr, $ARRAY_TYPE:ident, $DATA_TYPE:ident) => {{
+        let mut values = vec![];
+        for arr in $ARRAY.iter() {
+            if let Some(arr) = arr {
+                let arr = downcast_arg!(arr, $ARRAY_TYPE);
+                let sum = arr.values().iter().sum::<$DATA_TYPE>();
+                values.push(sum);
+            }
+        }
+        Ok(Arc::new($ARRAY_TYPE::from(values)))
+    }}
+);
+
+/// array_aggregate SQL function
+pub fn array_aggregate(args: &[ArrayRef]) -> Result<ArrayRef> {
+    assert_eq!(args.len(), 2);
+    let func_name = args[1].as_string::<i32>().value(0);
+    let args = &args[0..1];
+    match func_name {
+        "sum" => array_sum(args),
```

Comment:
If you implement this in terms of an [`AggregateExpr`](https://docs.rs/datafusion/latest/datafusion/physical_expr/trait.AggregateExpr.html) and then [Accumulator](https://docs.rs/datafusion/latest/datafusion/logical_expr/trait.Accumulator.html#) you could use the existing aggregate implementations. This would have several benefits:
1. We would avoid code duplication and get all existing aggregates "for free"
2. As new aggregates were added they would also be usable as an array aggregate
3. They would work for non primitive types (e.g. DecimalArray, strings, etc)

The basic idea would be to look up the [aggregate expr](https://docs.rs/datafusion/latest/datafusion/logical_expr/enum.Expr.html#variant.AggregateFunction) during the analysis / parsing phase, create a physical version, and then use it here to instantiate the accumulator

Maybe you could prototype this approach by simply hard coding the mapping `sum` --> [[`AggregateFunction::sum`]](https://docs.rs/datafusion/latest/datafusion/physical_expr/aggregate/build_in/enum.AggregateFunction.html) here and using that to create an `Accumulator` 🤔